File: rust/src/leetcode/editor/cn/minimum_depth_of_binary_tree.rs
use crate::leetcode::editor::cn::util::TreeNode;

struct Solution;
//leetcode submit region begin(Prohibit modification and deletion)
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }


use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn min_depth(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        return match root {
            None => 0,
            Some(node) => {
                let mut q = std::collections::VecDeque::new();
                q.reserve(10);
                q.push_back(node);
                let mut depth = 1;
                while !q.is_empty() {
                    let sz = q.len();
                    for _ in 0..sz {
                        let node = q.pop_front().unwrap();
                        let node = node.borrow();
                        if node.left.is_none() && node.right.is_none() {
                            return depth;
                        }
                        if node.left.is_some() {
                            q.push_back(node.left.as_ref().unwrap().clone());
                        }
                        if node.right.is_some() {
                            q.push_back(node.right.as_ref().unwrap().clone());
                        }
                    }

                    depth = depth + 1;
                }
                depth
            }
        };
    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: cpp/leetcodecn/minimum_depth_of_binary_tree.cpp
#include <vector>
#include <queue>
#include "util.h"

using namespace std;

//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int minDepth(TreeNode *root) {
        if (root == nullptr) return 0;

        queue<TreeNode *> q;
        q.push(root);

        int depth = 1;
        while (!q.empty()) {
            auto len = q.size();
            for (auto i = 0; i < len; i++) {
                auto front = q.front();
                q.pop();
                if (front->left == nullptr && front->right == nullptr) return depth;
                if (front->left != nullptr) q.push(front->left);
                if (front->right != nullptr) q.push(front->right);
            }
            depth++;
        }

        return depth;

    }
};
//leetcode submit region end(Prohibit modification and deletion)


int main(int argv, char *argc[]) {
}